typing/includecore.ml

```diff
@@ -372,11 +387,28 @@ and compare_records ~loc env params1 params2 n
 let compare_records_with_representation ~loc env params1 params2 n
       labels1 labels2 rep1 rep2
   =
-  match compare_records ~loc env params1 params2 n labels1 labels2 with
-  | None when rep1 <> rep2 ->
-      let pos = if rep2 = Record_float then Second else First in
-      Some (Unboxed_float_representation pos)
-  | err -> err
+  let err = compare_records ~loc env params1 params2 n labels1 labels2 in
+  match err, rep1, rep2 with
+  | None, Record_regular, Record_regular
+  | None, Record_float, Record_float
+  | None, Record_unboxed _, Record_unboxed _
+  | None, Record_inlined _, Record_inlined _
+  | None, Record_extension _, Record_extension _ ->
+     None
+  | Some err, _, _ ->
+     Some (Record_mismatch err)
+  | None, Record_unboxed _, _ ->
+     Some (Unboxed_representation First)
+  | None, _, Record_unboxed _ ->
+     Some (Unboxed_representation Second)
+  | None, Record_float, _ ->
+     Some (Record_mismatch (Unboxed_float_representation First))
+  | None, _, Record_float ->
+     Some (Record_mismatch (Unboxed_float_representation Second))
+  | None,
+    (Record_regular|Record_inlined _|Record_extension _),
+    (Record_regular|Record_inlined _|Record_extension _) ->
+     assert false
```

Comment:
a comment on why we `assert false` here could be nice

Reply:
I have no idea why we assert false here! In trunk at the moment, if you can manage to get this code to compare, say, a `Record_regular` and a `Record_inlined ...`, then it will error with a nonsensical message about unboxed floats. If this situation is possible, then an assertion failure is an improvement.